### PR TITLE
Support exchange of primitive messages with other MQTT clients

### PR DESCRIPTION
**Pull request by lreiher**
This PR integrates functionality for communicating with other (non-`mqtt_client`) MQTT clients. This feature allows ROS-based devices to exchange primitive messages with devices not based on ROS. It can be used in both directions:
1. Subscribe any (non-ROS-based) MQTT message and publish it as a primitive ROS message (`std_msgs/Bool`, `std_msgs/Int32`, `std_msgs/Float32`, `std_msgs/String`).
2. Subscribe any primitive ROS message  (`std_msgs/Bool`, `std_msgs/Int32`, `std_msgs/Float32`, `std_msgs/String`, ...) and publish it as a primitive MQTT string payload, ready to be interpreted by other MQTT clients.

---

***Excerpt from updated README***

As seen in the [Quick Start](#quick-start), the *mqtt_client* can not only exchange arbitrary ROS messages with other *mqtt_clients*, but it can also exchange primitive message data with other non-*mqtt_client* MQTT clients. The `primitive` parameter can be set for both ROS-to-MQTT (`bridge/ros2mqtt`) and MQTT-to-ROS (`bridge/mqtt2ros`) transmissions.

If a ROS-to-MQTT transmission is configured as `primitive`, the ROS message is simply serialized to a string representation, without providing any information of the underlying ROS message type via MQTT. If the ROS message type is one of the supported primitive ROS message types, the encapsualating ROS message components are also removed, s.t. only the raw data is published as a string. The supported primitive ROS message types are [`std_msgs/String`](http://docs.ros.org/en/api/std_msgs/html/msg/String.html), [`std_msgs/Bool`](http://docs.ros.org/en/api/std_msgs/html/msg/Bool.html), [`std_msgs/Char`](http://docs.ros.org/en/api/std_msgs/html/msg/Char.html), [`std_msgs/UInt8`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt8.html), [`std_msgs/UInt16`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt16.html), [`std_msgs/UInt32`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt32.html), [`std_msgs/UInt64`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt16.html), [`std_msgs/Int8`](http://docs.ros.org/en/api/std_msgs/html/msg/Int8.html), [`std_msgs/Int16`](http://docs.ros.org/en/api/std_msgs/html/msg/Int16.html), [`std_msgs/Int32`](http://docs.ros.org/en/api/std_msgs/html/msg/Int32.html), [`std_msgs/Int64`](http://docs.ros.org/en/api/std_msgs/html/msg/Int64.html), [`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float32.html), [`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float64.html).

If an MQTT-to-ROS transmission is configured as `primitive`, the MQTT message is interpreted and published as a primitive data type, if possible. The message is probed in the following order: `bool` ([`std_msgs/Bool`](http://docs.ros.org/en/api/std_msgs/html/msg/Bool.html)), `int` ([`std_msgs/Int32`](http://docs.ros.org/en/api/std_msgs/html/msg/Int32.html)), `float` ([`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float32.html)), `string` ([`std_msgs/String`](http://docs.ros.org/en/api/std_msgs/html/msg/String.html)).

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -196,6 +226,14 @@ bridge:
...
...orted primitive ROS message types, the encapsualating ROS message components are also removed, s.t. only the raw data is published as a string. The supported primitive ROS message types are [`std_msgs/String`](http://docs.ros.org/en/api/std_msgs/html/msg/String.html), [`std_msgs/Bool`](http://docs.ros.org/en/api/std_msgs/html/msg/Bool.html), [`std_msgs/Char`](http://docs.ros.org/en/api/std_msgs/html/msg/Char.html), [`std_msgs/UInt8`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt8.html), [`std_msgs/UInt16`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt16.html), [`std_msgs/UInt32`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt32.html), [`std_msgs/UInt64`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt16.html), [`std_msgs/Int8`](http://docs.ros.org/en/api/std_msgs/html/msg/Int8.html), [`std_msgs/Int16`](http://docs.ros.org/en/api/std_msgs/html/msg/Int16.html), [`std_msgs/Int32`](http://docs.ros.org/en/api/std_msgs/html/msg/Int32.html), [`std_msgs/Int64`](http://docs.ros.org/en/api/std_msgs/html/msg/Int64.html), [`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float32.html), [`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float64.html).
```
Comment (bastilam) [inline README.md:233]:
```suggestion
If a ROS-to-MQTT transmission is configured as `primitive`, the ROS message is simply serialized to a string representation, without providing any information on the underlying ROS message type via MQTT. If the ROS message type is one of the supported primitive ROS message types, the encapsulating ROS message components are also removed, s.t. only the raw data is published as a string. The supported primitive ROS message types are [`std_msgs/String`](http://docs.ros.org/en/api/std_msgs/html/msg/String.html), [`std_msgs/Bool`](http://docs.ros.org/en/api/std_msgs/html/msg/Bool.html), [`std_msgs/Char`](http://docs.ros.org/en/api/std_msgs/html/msg/Char.html), [`std_msgs/UInt8`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt8.html), [`std_msgs/UInt16`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt16.html), [`std_msgs/UInt32`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt32.html), [`std_msgs/UInt64`](http://docs.ros.org/en/api/std_msgs/html/msg/UInt16.html), [`std_msgs/Int8`](http://docs.ros.org/en/api/std_msgs/html/msg/Int8.html), [`std_msgs/Int16`](http://docs.ros.org/en/api/std_msgs/html/msg/Int16.html), [`std_msgs/Int32`](http://docs.ros.org/en/api/std_msgs/html/msg/Int32.html), [`std_msgs/Int64`](http://docs.ros.org/en/api/std_msgs/html/msg/Int64.html), [`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float32.html), [`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float64.html).
```

File: src/MqttClient.cpp [hunk 2]
```diff
@@ -133,9 +228,20 @@ void MqttClient::loadParameters() {
...
+          // ros2mqtt[k]/primitive
+          if (ros2mqtt_params[k].hasMember("primitive"))
+            ros2mqtt.primitive = ros2mqtt_params[k]["primitive"];
+
           // ros2mqtt[k]/inject_timestamp
           if (ros2mqtt_params[k].hasMember("inject_timestamp"))
             ros2mqtt.stamped = ros2mqtt_params[k]["inject_timestamp"];
+          if (ros2mqtt.stamped && ros2mqtt.primitive) {
+            NODELET_WARN(
+              "Timestamp will not be injected into primitive messages on ROS "
```
Comment (bastilam) [inline src/MqttClient.cpp:240]: Add hint to Readme that timestamping is not possible with primitive messages?

**Review by bastilam — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -178,6 +206,7 @@ bridge:
   ros2mqtt:            # array specifying which ROS topics to map to which MQTT topics
     - ros_topic:         # ROS topic whose messages are transformed to MQTT messages
       mqtt_topic:        # MQTT topic on which the corresponding ROS messages are sent to the broker
+      primitive:         # [false] whether to publish as primitive message
```
[the remark this answers is not in the document]
Comment (lreiher) [reply README.md:209]: Can already be realized by specifying two bridges with the same ROS topic.

File: README.md [hunk 4]
```diff
@@ -196,6 +226,14 @@ bridge:
...
+
+If an MQTT-to-ROS transmission is configured as `primitive`, the MQTT message is interpreted and published as a primitive data type, if possible. The message is probed in the following order: `bool` ([`std_msgs/Bool`](http://docs.ros.org/en/api/std_msgs/html/msg/Bool.html)), `int` ([`std_msgs/Int32`](http://docs.ros.org/en/api/std_msgs/html/msg/Int32.html)), `float` ([`std_msgs/Float32`](http://docs.ros.org/en/api/std_msgs/html/msg/Float32.html)), `string` ([`std_msgs/String`](http://docs.ros.org/en/api/std_msgs/html/msg/String.html)).
+
 ## Latency Computation
 
 The *mqtt_client* provides built-in functionality to measure the latency of transferring a ROS message via an MQTT broker back to ROS. To this end, the sending client injects the current timestamp into the MQTT message. The receiving client can then compute the latency between message reception time and the injected timestamp. **Naturally, this is only accurate to the level of synchronization between clocks on sending and receiving machine.**
```
Comment (lreiher) [inline README.md:239]:
```suggestion
The *mqtt_client* provides built-in functionality to measure the latency of transferring a ROS message via an MQTT broker back to ROS. Note that this functionality is only available for non-primitive messages (see [Primitive Messages](#primitive-messages)). To this end, the sending client injects the current timestamp into the MQTT message. The receiving client can then compute the latency between message reception time and the injected timestamp. **Naturally, this is only accurate to the level of synchronization between clocks on sending and receiving machine.**
```

File: README.md [hunk 5]
```diff
@@ -5,7 +5,7 @@
   <img src="https://img.shields.io/github/v/release/ika-rwth-aachen/mqtt_client"/></a>
   <img src="https://img.shields.io/github/license/ika-rwth-aachen/mqtt_client"/></a>
   <a href="https://github.com/ika-rwth-aachen/mqtt_client/actions/workflows/build.yml"><img src="https://github.com/ika-rwth-aachen/mqtt_client/actions/workflows/build.yml/badge.svg"/></a>
-  <img src="https://img.shields.io/github/stars/ika-rwth-aachen/mqtt_client?style=social"/></a>
+  <a href="https://github.com/ika-rwth-aachen/mqtt_client"><img src="https://img.shields.io/github/stars/ika-rwth-aachen/mqtt_client?style=social"/></a>
 </p>
 
 The *mqtt_client* package provides a ROS nodelet that enables connected ROS-based devices or robots to exchange ROS messages via an MQTT broker using the [MQTT](http://mqtt.org) protocol. This works generically for arbitrary ROS message types.
```
Comment (bastilam) [inline README.md:11]: I would suggest to make this information more prominent.
```suggestion
The *mqtt_client* package provides a ROS nodelet that enables connected ROS-based devices or robots to exchange ROS messages via an MQTT broker using the [MQTT](http://mqtt.org) protocol. This works generically for arbitrary ROS message types. The *mqtt_client* can also exchange primitive messages with MQTT clients running on devices not based on ROS.
```

File: README.md [hunk 6]
```diff
@@ -82,33 +91,52 @@ roslaunch mqtt_client standalone.launch
...
+```
+
+```bash
+# 4th terminal: publish primitive ROS message to /ping/primitive
+rostopic pub -r 1 /ping/primitive std_msgs/Int32 42
+```
+
+```bash
+# 5th terminal: publish primitive MQTT message to pingpong/primitive
+docker run --rm --network host eclipse-mosquitto mosquitto_pub -h localhost -t "pingpong/primitive" --repeat 10 --repeat-delay 1 -m 69
```
Comment (bastilam) [inline README.md:136]: ROS messages are sent indefinitely, primitive messages are only sent 10 times. I did not find a good way to send the primitive messages indefinitely but maybe we can increase this to 100 to give enough time to find the correct terminal again ;-)

```suggestion
docker run --rm --network host eclipse-mosquitto mosquitto_pub -h localhost -t "pingpong/primitive" --repeat 100 --repeat-delay 1 -m 69
```

File: README.md [hunk 7]
```diff
@@ -82,33 +91,52 @@ roslaunch mqtt_client standalone.launch
...
 
 Note that the *mqtt_client* successfully connected to the broker and also echoed which ROS/MQTT topics are being bridged.
 
-In order to test the communication, publish any message on ROS topic `/ping` and wait for a response on ROS topic `/pong`. To this end, open two new terminals and execute the following commands.
+In order to test the communication among *mqtt_clients*, publish any ROS message on ROS topic `/ping/ros` and wait for a response on ROS topic `/pong/ros`.
 
 ```bash
-# 1st terminal: listen on /pong
-rostopic echo /pong
+# 1st terminal: listen for ROS messages on /pong/ros
```
Comment (bastilam) [inline README.md:113]: The numbering of terminals in the quick start might be misleading. 

`docker run --rm --network host --name mosquitto eclipse-mosquitto` is the first terminal in the quick start.

`roslaunch mqtt_client standalone.launch` is the second terminal in the quick start.

So here, we would start our 3rd terminal and the next terminal numbers would need to be increased as well.

**Review by bastilam — CHANGES_REQUESTED** [review]

File: README.md [hunk 8]
```diff
@@ -82,33 +91,52 @@ roslaunch mqtt_client standalone.launch
...
+[ WARN] [1665575657.360847295]: Parameter 'client/max_inflight' not set, defaulting to '65535'
+[ INFO] [1665575657.361281461]: Bridging ROS topic '/ping/ros' to MQTT topic 'pingpong/ros' 
+[ INFO] [1665575657.361303380]: Bridging primitive ROS topic '/ping/primitive' to MQTT topic 'pingpong/primitive' 
+[ INFO] [1665575657.361352809]: Bridging MQTT topic 'pingpong/ros' to ROS topic '/pong/ros'
+[ INFO] [1665575657.361370558]: Bridging MQTT topic 'pingpong/primitive' to primitive ROS topic '/pong/primitive'
+[ INFO] [1665575657.362153083]: Connecting to broker at 'tcp://localhost:1883' ...
+[ INFO] [1665575657.462622065]: Connected to broker at 'tcp://localhost:1883'
 ```
 
 Note that the *mqtt_client* successfully connected to the broker and also echoed which ROS/MQTT topics are being bridged.
```
Comment (lreiher) [inline README.md:108]:
```suggestion
Note that the *mqtt_client* successfully connected to the broker and also echoed which ROS/MQTT topics are being bridged. For testing the communication between *mqtt_client*, itself, and other MQTT clients, open five new terminals.
```

File: README.md (hunk 7 above)
Comment (lreiher) [reply README.md:113]: Added sentence to open 5 new terminals, to which these numbers are then referring to.

**Pull request merged**